tools/testing/e2e/iree-e2e-attention-test.cc

```diff
@@ -0,0 +1,476 @@
+  }
+  if (iree_status_is_ok(status)) {
+    out_results->actual_contents.data_length =
+        iree_hal_buffer_byte_length(result_buffer);
+    status = iree_allocator_malloc(host_allocator,
+                                   out_results->actual_contents.data_length,
+                                   (void**)&out_results->actual_contents.data);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_device_transfer_d2h(
+        device, result_buffer, 0, out_results->actual_contents.data,
+        out_results->actual_contents.data_length,
+        IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT, iree_infinite_timeout());
+  }
+  if (iree_status_is_ok(status)) {
+    out_results->expected_contents.data_length =
+        iree_hal_buffer_byte_length(result_buffer);
+    status = iree_allocator_malloc(
+        host_allocator, out_results->expected_contents.data_length,
+        (void**)&out_results->expected_contents.data);
+  }
+  if (!iree_status_is_ok(status)) {
+    attention_results_deinitialize(out_results);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void attention_results_deinitialize(attention_results_t* results) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_allocator_free(results->host_allocator, results->query_contents.data);
+  iree_allocator_free(results->host_allocator, results->key_contents.data);
+  iree_allocator_free(results->host_allocator, results->value_contents.data);
+  IREE_TRACE_ZONE_END(z0);
+}
+
+// Helper for check_matmul_results: the actual interesting part once we've
+// obtained and validated the {m,k,n}_size values. On error, detailed logging is
+// written to |file| if it is not NULL.
+static iree_status_t check_attention_results_impl(
+    FILE* file, const attention_results_t* results, int check_every) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, reference_attention(results->b, results->m, results->k1, results->k2,
+                              results->n, results->query_elem_type,
+                              results->key_elem_type, results->value_elem_type,
+                              results->query_contents, results->key_contents,
+                              results->value_contents, results->actual_contents,
+                              results->expected_contents, check_every));
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+// Given an actual attention's inputs and output (all host-local), uses a
+// reference attention implementation on the same inputs to check if the output
+// is correct. On error, detailed logging is written to |file| if it is not
+// NULL.
+static iree_status_t check_attention_results(
+    FILE* file, const attention_results_t* results) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  // TODO: Increase the check every param to reduce the number of comparisons.
+  int check_every = 1;
+  iree_status_t status =
+      check_attention_results_impl(file, results, check_every);
```

Comment:
How much data are we talking about here? How many elements for the different test cases?

Reply:
What do those translate to in terms of bytes? (I'm wondering if we can store inputs and outputs on disk somewhere, or if we _need_ to generate them using a reference implementation at test time)

Reply:
Should be like this:
Codegen:
2(since FP16) x (2 x 512 x 64 x 128 x 32) --> 536870912 bytes
2(since FP16) x (2 x1024x128x256 x 64) --> 8589934592 bytes
2(since FP16) x (2 x 2048 x 256 x 512 x 128) --> 137438953472 bytes
Reference:
4(since FP32) x (2 x 512 x 64 x 128 x 32)  --> 1073741824 bytes
4(since FP32) x (2 x1024x128x256 x 64) --> 17179869184 bytes
4(since FP32) x (2 x 2048 x 256 x 512 x 128) --> 274877906944 bytes

Reply:
Really? 274 gigabytes? Is that all input/output data, or just some intermediate buffer?

Reply:
I updated the calculation above.

Reply:
Oh. 4 _megabytes_ is quite reasonable to store in git LFS, especially with a local cache. 1000 test cases would be 4GB though, and GitHub has hilariously low storage and bandwidth quotas of 1GB/month: https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-storage-and-bandwidth-usage

Reply:
FWIW, based on attention experiments, random data for attention accuracy tests generally works well and gets out accuracy issues, as long as you are covering the entire precision range.